### PR TITLE
Add support for compositing using Metal on macOS

### PR DESCRIPTION
**Pull request by gw280**
This adds support for Metal compositing on macOS.

https://github.com/flutter/flutter/issues/81320

Tests to come.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

### CODE REVIEW

**Comment by gw280** [comment]
FlutterMetalTextureProvider may be unnecessary, we can probably just throw that code inline into the callsite. I do want to clean up FlutterFrameBufferProvider though as the name "Provider" implies it's a factory and not a wrapper that manages the FBO for us.

File: shell/platform/darwin/macos/framework/Source/FlutterBackingStoreData.mm [hunk 1]
```diff
@@ -12,11 +12,13 @@
 @implementation FlutterBackingStoreData
 
 - (nullable instancetype)initWithLayerId:(size_t)layerId
-                              fbProvider:(nonnull FlutterFrameBufferProvider*)fbProvider
+                              fbProvider:(nullable FlutterFrameBufferProvider*)fbProvider
                          ioSurfaceHolder:(nonnull FlutterIOSurfaceHolder*)ioSurfaceHolder {
   if (self = [super init]) {
     _layerId = layerId;
-    _frameBufferProvider = fbProvider;
```
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterBackingStoreData.mm:19]: why does this matter? `_frameBufferProvider` can always be `fbProvider` without the null check right?

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.h [hunk 2]
```diff
@@ -52,9 +53,30 @@ class FlutterCompositor {
...
   PresentCallback present_callback_;
 
+  // frame_started_ keeps track of if a layer has been
+  // created for the frame.
+  bool frame_started_ = false;
+
+  // Count for how many CALayers have been created for a frame.
+  // Resets when a frame is finished.
+  // ca_layer_count_ is also used as a layerId.
+  size_t ca_layer_count_ = 0;
```
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:65]: will this always be equal to `ca_layer_map_.size()`?

File: shell/platform/darwin/macos/framework/Source/FlutterMetalCompositor.mm [hunk 3]
```diff
@@ -3,27 +3,99 @@
...
+    // FlutterView.
+    FlutterMetalRenderBackingStore* backingStore =
+        reinterpret_cast<FlutterMetalRenderBackingStore*>(
+            [view_controller_.flutterView backingStoreForSize:size]);
+    backing_store_out->metal.texture.texture =
+        (__bridge FlutterMetalTextureHandle)backingStore.texture;
+  } else {
+    FlutterMetalTextureProvider* textureProvider =
+        [[FlutterMetalTextureProvider alloc] initWithMTLDevice:mtl_device_];
+    FlutterIOSurfaceHolder* io_surface_holder = [FlutterIOSurfaceHolder alloc];
```
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterMetalCompositor.mm:41]: `alloc] init]`

File: shell/platform/darwin/macos/framework/Source/FlutterMetalTextureProvider.mm [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+@implementation FlutterMetalTextureProvider
+- (instancetype)initWithMTLDevice:(const id<MTLDevice>)device {
+  if (self = [super init]) {
+    _device = device;
+  }
+  return self;
+}
+
+- (id<MTLTexture>)createTextureWithSize:(CGSize)size iosurface:(nonnull IOSurfaceRef) ioSurface {
+  auto texture_descriptor = [MTLTextureDescriptor texture2DDescriptorWithPixelFormat:MTLPixelFormatBGRA8Unorm
```
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterMetalTextureProvider.mm:23]: we have similar functionality in `FlutterSurfaceManager` as well, can we unify?

**Review by iskakaushik — COMMENTED** [review]
looking good so far

File: shell/platform/darwin/macos/framework/Source/FlutterBackingStoreData.mm (hunk 1 above)
Comment (gw280) [reply shell/platform/darwin/macos/framework/Source/FlutterBackingStoreData.mm:19]: yeah it doesn't matter, just habit.

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.h (hunk 2 above)
Comment (gw280) [reply shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:65]: This is just refactored existing code that I moved from FlutterGLCompositor into FlutterCompositor. A closer look at it suggests that ca_layer_count_ is always going to be the size of the map; that being said, it's concerning to me that this map expands in size infinitely and is never reduced in size. I think there's a separate bug here where we need to figure out when to purge old items from the map.

File: shell/platform/darwin/macos/framework/Source/FlutterMetalTextureProvider.mm (hunk 4 above)
Comment (gw280) [reply shell/platform/darwin/macos/framework/Source/FlutterMetalTextureProvider.mm:23]: I was thinking some more about this. I'm not sure it's worth unifying FlutterSurfaceManager with this as it's literally only two lines of code, but I think I will get rid of FlutterMetalTextureProvider and just make the texture in FlutterMetalCompositor. Two lines of code doesn't seem worth the complexity of refactoring.

**Comment by gw280** [comment]
There's something wrong with this, I think it's working by accident. Going to close for now, will open a new PR when I have it working properly.

**Pull request closed without merging**